src/renderer/components/RecentProjectsBlock.tsx

```diff
@@ -70,6 +74,25 @@ const RecentProjectsBlock = () => {
     .sort(sortByDateModified)
     .slice(0, RECENT_PROJECTS_COUNT);
 
+  // TODO(chloe) I think this can cause a memory leak if the request is in flight
+  // when the page changes; not sure how to fix
+  useEffect(() => {
+    Promise.all(
+      recentProjects.map(async ({ id }) => ({
+        id,
+        thumbnail: await ipc.loadThumbnail(id),
+      }))
+    )
+      .then((list) => {
+        const map: Record<string, string> = {};
+        list.forEach(({ id, thumbnail }) => {
+          map[id] = thumbnail;
+        });
+        setThumbnails(map);
+      })
+      .catch();
+  }, [recentProjects]);
```

Comment:
Ok yes there will be a memory leak. There needs to be a cleanup function for the useEffect (basically the useEffect should return a function than cancels any subscriptions in it) that aborts the ipc request. In a fetch request equivalent I would abort the fetch call using an abortController and then the catch would catch the AbortError created. 

However, I'm struggling to figure out a way to abort an IPC request. Initial research indicates that it is not possible. We can come back to this.

The more pressing issue I feel is that the memory leak is apparent when testing with 1 recent project. This shouldn't happen as the user opening a recent project should happen well after all of the thumbnails have been loaded (it's not a huge process I think).

If you chuck a `console.log` in the useEffect you'll see that `recentProjects` is getting updated constantly. Again, this should not be happening. The problem is that the state variable `recentProjects` is being modified directly (sorted and filtered). I don't think this is the correct pattern.

If we want to always have access to the filtered / sorted version of recent projects then that should be stored in another state variable (eg: `recentProjectsSortedAndFiltered` which is updated (via another useEffect) whenever the original state variable (`recentProjects`) is modified.

![image](https://user-images.githubusercontent.com/101784287/182758968-9a978092-a6b1-407b-bd12-a38c813f84ba.png)

In the example above, 'dog' should only get printed whenever `recentProjects` is modified or on page unmounted.

![image](https://user-images.githubusercontent.com/101784287/182759106-2baae56b-6239-4c39-abcc-2b22cea94f54.png)

Reply:
In regards to figuring out a way to abort the IPC request, I really don't know.

While the memory leak could be a problem, it will not be common as the user would need to have quite a few projects. While, I would like to address it, I think the time required (I already spent like 20 mins) would outweigh the reliability and performance benefits.

If you disagree, I'm happy to discuss in architecture meeting.

Reply:
Seems reasonable. As for the re-rendering, we can use `useMemo` instead of `useEffect` for the sorted + sliced list as it's a bit more idiomatic. Trying now

Reply:
Regarding aborting the IPC I have two ideas:
1. find a generic "cancelable promise" library and use that - may not be possible
2. add an `AsyncState` to the thumbnail loading request and prevent page changes if that request is in flight (thereby stopping the component from unmounting)

I still think it's probably not worth the effort to do either of these, clearly any performance issues from the bug were caused by the 'constant rerender' issue which is now fixed. Good catch on that btw, I think I realised when I wrote it it was lazy but figured it wouldn't be an issue since at that time there was no cause for the home screen state to change

Reply:
useMemo! thanks for taking it one step further :)

I like the async state idea but agree not worth the time. We can always implement it later if we come across a case where aborting an IPC request will have a significant performance impact (perhaps transcription???)